Fixed a bug where invalidated worker resource caches are not GC-ed

evanchaoli opened the pull request:
## Changes proposed by this PR

closes #8473 


* [x] db schema change (up)
* [x] db schema change (down)
* [x] update `db.WorkerResourceCaches` 
* [x] update `InitResourceCache` and `InitStreamedResourceCache`
* [x] update unit tests
* [x] add a GC component to clean up invalidated worker resource caches 
* [x] run end-to-end tests   
* [x] done

In summary, the solution is:

* Add `where` clause to unique index `worker_resource_caches_uniq`, so that invalidated caches will not conflict to valid caches.
* Add `invalid_since` column to table `worker_resource_caches`, it will be auto set by a db trigger, so that GC component may delete invalidated caches based on time.
* Updated `db.WorkerResourceCaches`, so that for `worker name` + `resource cache id`, there can be only one valid cache, and multiple invalidated caches.
* If a build started before `invalid_since`, then the cache can still be used by the build; otherwise the build cannot use the cache.
* Add a GC component to delete invalidated worker resource caches based on `invalid_since` time.


## Notes to reviewer

<!--
If needed, leave any special pointers for reviewing or testing your PR.
-->

## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next Concourse release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Introduce new pipeline UI algorithm

Example notes:

* Reticulating splines is the new process Concourse uses to create the network
  of lines between jobs.
* Combines many short lines and curves into a network of splines.

If there are no additional notes necessary you may remove this entire section.
-->

* <!-- remove if no additional notes needed -->

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

evanchaoli committed:
WIP: fixed a bug where invalidated worker resource caches are not GC-ed

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
fix unit tests

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
fix unit tests

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
add db migration down script

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
removed FIt.

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
WIP

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
big refactor

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
add worker-resoource-cache gc

Signed-off-by: Evan <chaol@vmware.com>

xtremerui reviewed (commented):
Just couple questions

evanchaoli commented on atc/worker/gardenruntime/worker.go:
```diff
@@ -715,6 +715,12 @@ func (worker *Worker) findOrStreamVolume(
+			// From `get` to where it artificat is used, it may take time, and during
+			// the period, original worker may have been pruned, but the artifact might
+			// have been streamed to other workers, thus find the artifact again can make
+			// sure the artifact exist.
```
With my latest commit, this TODO is no longer needed. Even if original worker is pruned, worker resource cache will be marked as invalidated, but it can still be used by running builds.

But if the source worker is pruned, then the build may still fail, which is a general problem of volume streaming, I am not going to solve the problem in this PR.

evanchaoli commented on atc/db/volume.go:
```diff
@@ -462,25 +472,26 @@ func (volume *createdVolume) initializeResourceCache(tx Tx, resourceCache Resour
+		return nil, false, ErrVolumeMissing
 	}
 
+	volume.workerResourceCacheID = workerResourceCache.ID
```
Before this PR, it uses `resource cache id` + `worker name` to look for `worker resource cache`, it supposed to have a single tuple found by the criteria. Now, that criteria cannot uniquely identify a worker resource cache, because it may be 1 valid cache and multiple invalidated caches. Thus I changed to use  `workerResourceCache.ID` to identify a specific cache.

evanchaoli committed:
addressed review comments

Signed-off-by: Evan <chaol@vmware.com>

Update atc/db/worker_resource_cache.go

Co-authored-by: Rui Yang <ruiy@outlook.com>

Update atc/db/worker_resource_caches_test.go

Co-authored-by: Rui Yang <ruiy@outlook.com>

Update atc/db/worker_resource_caches_test.go

Co-authored-by: Rui Yang <ruiy@outlook.com>

xtremerui commented on atc/db/worker_resource_cache.go:
```diff
@@ -79,38 +89,101 @@ func (workerResourceCache WorkerResourceCache) FindOrCreate(tx Tx, sourceWorkerB
 }
 
-func (workerResourceCache WorkerResourceCache) find(runner sq.Runner) (*UsedWorkerResourceCache, bool, error) {
+// find should only return a valid worker resource cache if volumeShouldBeValidBefore is nil.
+// When volumeShouldBeValidBefore is not nil, if a cache's invalid_since is later than
+// volumeShouldBeValidBefore, then the cache is considered as valid. If there is a valid cache,
+// then the valid cache should be returned; if there are multiple invalid caches with
+// invalid_since later than volumeShouldBeValidBefore, the newest invalid_since one should be
+// returned.
```
```suggestion
// find should return a valid worker resource cache if 
//   volumeShouldBeValidBefore is nil OR
//   a cache's invalid_since is later than volumeShouldBeValidBefore.
// If there are multiple invalid caches with invalid_since later than volumeShouldBeValidBefore, 
// the cache with newest invalid_since should be returned.
```

xtremerui replied:
@evanchaoli looks like the comment here is changed back to the old version

xtremerui commented on atc/db/worker_resource_caches_test.go:
```diff
@@ -0,0 +1,240 @@
+			})
+
+			Context("Create a worker resource cache again on worker0 again with it's own base resource type", func() {
+				var uwrc2 *db.UsedWorkerResourceCache
```
```suggestion
				var uwrc *db.UsedWorkerResourceCache
```

xtremerui commented on atc/db/worker_resource_caches_test.go:
```diff
@@ -0,0 +1,240 @@
+			})
+
+			Context("Create a worker resource cache again on worker0 with worker1's base resource type", func() {
+				var uwrc2 *db.UsedWorkerResourceCache
```
```suggestion
				var uwrc *db.UsedWorkerResourceCache
```

xtremerui commented on atc/db/worker_resource_caches_test.go:
```diff
@@ -0,0 +1,240 @@
+			})
+
+			Context("Create a worker resource cache on worker1 with worker0's base base resource type", func(){
+				var uwrc2 *db.UsedWorkerResourceCache
```
```suggestion
				var uwrc *db.UsedWorkerResourceCache
```

xtremerui reviewed (commented):
I'd like to make the docs and tests more clear and readable.

evanchaoli committed:
address review comments

Signed-off-by: Evan <chaol@vmware.com>

xtremerui reviewed (approved):
LGTM!

evanchaoli commented:
@xtremerui Please hold merge this PR. We have tried to deployed this PR to our prod cluster and found it might cause ATC CPU usages to increase ~10%. I will enhance this PR to solve the CPU usage increasing problem.

evanchaoli committed:
optimized implementation of WorkerResourceCache.find().

Signed-off-by: Evan <chaol@vmware.com>

deepakmohanakrishnan07 commented on atc/db/worker_resource_cache.go:
```diff
@@ -124,65 +124,64 @@ func (workerResourceCache WorkerResourceCache) FindByID(runner sq.Runner, id int
+				 valid_caches AS (
+                     SELECT id, worker_base_resource_type_id, invalid_since, 1 as priority
+                       FROM candidates
+                       WHERE worker_base_resource_type_id is not null
```
This is great, I have been following this change in the recent past to overcome this issue in our production cluster. 

I have a quick question on this psql query. IIUC, Does this WHERE clause check for `worker_base_resource_type_id != 0`? Because, I think the `worker_base_resource_type_id` is set to `0` when a worker is either pruned or the cache is invalidated.

evanchaoli replied:
@deepakmohanakrishnan07 No, `worker_base_resource_type_id` is set to NULL. You may check with your db to verify that.

xtremerui commented:
@evanchaoli just a follow up on this issue as we are now in progress of releasing 7.9.

evanchaoli commented:
@xtremerui We are planning to apply this patch to stage this week and prod next week. So if 7.9 is planned for Nov., there should be enough time to prove this patch in our clusters.

evanchaoli committed:
fixed "for share" bug in WorkerResourceCache.find().

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
Enhanced resourceCacheLifecycle.CleanInvalidWorkerResourceCaches().

Signed-off-by: Evan <chaol@vmware.com>

evanchaoli committed:
Bump db migration script's timestamp.

Signed-off-by: Evan <chaol@vmware.com>
